### ISSUE TITLE
NoiseTexture2D seamless not working

### ISSUE THREAD

**Issue by beev1s** [godotengine/godot#71996]
### Godot version

Godot 4.0 Beta 15

### System information

Windows 10

### Issue description

Scrolling a seamless 2D noise texture in 2D does not work as the texture does not appear to be seamless.
The visual shader is the same between Godot 3.5.1.stable and Godot 4.0 Beta 15.
The NoiseTexture is supposed to be moved by 0.5 in all examples.

![BrokenScrollingNoiseTexture](https://user-images.githubusercontent.com/38868780/214396381-5a8e0097-d49a-4cff-b61f-9ed30ce23df9.PNG)
Broken in 2D

![ScrollingNoiseTextureIn3D](https://user-images.githubusercontent.com/38868780/214396372-294e2e08-cad5-4880-9d21-24d7f9b6355d.PNG)
Works in 3D on a planar mesh

![WorkingScrollingNoiseTexture](https://user-images.githubusercontent.com/38868780/214396394-be3b6391-b5e7-42d5-964f-bf36f3f3608a.PNG)
Works in Godot 3.5.1.stable

### Steps to reproduce

Open Scrolling2DNoiseTexture.tscn in the appended example

### Minimal reproduction project

[Broken2dSeamlessNoiseTextures.zip](https://github.com/godotengine/godot/files/10493517/Broken2dSeamlessNoiseTextures.zip)

**Comment by clayjohn** [comment]
Do you have repeat enabled on your CanvasItem?

**Comment by beev1s** [comment]
I did not not, though enabling it did not change anything

**Comment by clayjohn** [comment]
It looks like the repeat property is not getting correctly set in the VisualShader then. This isn't an issue with the NoiseTexture itself as the repeat mode is set within the shader.

**Comment by aXu-AP** [comment]
Also doesn't work using `Texture2DParameter`, which gives more options. I presume setting `repeat` to `default` would get it's value from `CanvasItem.texture_repeat` or project setting `rendering/textures/canvas_textures/default_texture_repeat`, but it seems to be disabled even if those are enabled.
However, setting `Texture2D` as texture gets right values from the node.
In the example below, noise texture is set as polygon's texture, and it repeats as it was set in node's properties. The icon texture however, doesn't repeat unless explicitly set to repeat from the shader itself.

![visual_shader](https://user-images.githubusercontent.com/1621768/214629422-909cc332-bfb5-4087-87bd-e3f9eb33e701.png)


This behaviour isn't only in visual shaders. In glsl shader the difference is following:

```
uniform sampler2D tex1 : repeat_disable; # Explicitly disabled
uniform sampler2D tex2; # By default disabled anyway
```
In 3.x this doesn't cause problems since repeat is set by texture basis on importing. In 4.x is there any way to change filter & repeat without changing shader code itself?

**Comment by KoBeWi** [comment]
I think this is intended behavior. The filter/repeat in CanvasItems apply only to the node's own texture (e.g. `texture` in Sprite2D). For shader uniforms you need to use the hint.

**Comment by aXu-AP** [comment]
Exactly, this was my understanding as well. However, I think it is an oversight since there's no way to change filtering/repeat flags from the editor without touching shader code. In a sense, it is a regression since in 3.x you could do it by changing texture import settings.

**Comment by miyanokomiya** [comment]
Workaround for someone in the future

![image](https://user-images.githubusercontent.com/20733354/229258722-2b420fe5-1ce1-4dbb-a09f-fcaf94623d02.png)

**Comment by Chaosus** [comment]
To fix it, you can enable the repeat mode on the Texture Parameter node:

![image](https://user-images.githubusercontent.com/3036176/229269718-3a8051d4-a4e1-4dfe-bfb6-6b2efab73d46.png)

**Comment by miyanokomiya** [comment]
I found repeat hint. That's better than https://github.com/godotengine/godot/issues/71996#issuecomment-1492774431
`uniform sampler2D noise: repeat_enable;`

**Comment by Chaosus** [comment]
> I found repeat hint. That's better than [#71996 (comment)](https://github.com/godotengine/godot/issues/71996#issuecomment-1492774431) `uniform sampler2D noise: repeat_enable;`

This issue is about visual shader and not about common shaders, and that flag is already been used by that check box.

**Comment by DasGandlaf** [comment]
To reiterate, there is some bug going on, afaik.
@Chaosus You've set repeat enabled for the texture in the scene. How can you do so, for a texture which only exists in the shader? Like so: (**repeat is enabled in project settings**):
![image](https://user-images.githubusercontent.com/63406944/233759097-4939a4c2-c436-431e-bc9a-bbfd980f3b3c.png)
Once you add texture 2d parameters:
![image](https://user-images.githubusercontent.com/63406944/233759179-c0acda66-7119-4648-ba8e-a54c06257fcc.png)

Did anyone find a fix, even if it's just a workaround?

**Comment by Chaosus** [comment]
@DasGandlaf If you connect the sampler port to the Texture2D you should also change the source to SamplerPort (red warning at the bottom of the node is showing if you forgot to do it)...

**Comment by DasGandlaf** [comment]
@Chaosus Yes, but changing it to sampler port, uses the texture from the scene. The one that was defined in the node.

**Comment by Chaosus** [comment]
@DasGandlaf Then simply replace that texture to the NoiseTexture in the inspector.

**Comment by DasGandlaf** [comment]
I don't think that is a fix. It wouldn't work, If I need two different scrolling, repeating noise textures for example, as I can only set one to the node in the inspector. It should allow to set `repeat enable`, to any texture that is in the shader, not only the one in the scene.

**Comment by kleonc** [comment]
> I hope this demonstrates the problem correctly: ![image](https://user-images.githubusercontent.com/63406944/233760086-9f993a78-a71b-40ed-a7cf-3a3eb6d5e501.png)
> 
> Using SamplerPort as the source, it uses the texture from the scene. But I want to set repeat on the noise, which I have imported from the FileSystem.

@DasGandlaf You've not connected it to the output color, that's why the default texture is shown. After adding a Texture2DParameter node you should be able to assign a texture to it through the inspector (be sure to save the shader after adding such node):
![Godot_v4 0 2-stable_win64_Ca82d6ErdF](https://user-images.githubusercontent.com/9283098/233775932-264c1ab2-2a48-46e3-9d12-7c144597fa62.png)![FMfUl0f8wI](https://user-images.githubusercontent.com/9283098/233775934-20e258db-b373-48df-a52c-e3fb2d874aa6.png)

Note the texture I've set is seamless:
![obraz](https://user-images.githubusercontent.com/9283098/233775993-01ff16b8-8ca5-4daa-97f2-6edd6be2aae7.png)

**Comment by DasGandlaf** [comment]
@kleonc Thank you, I appreciate you trying to help, but I'm afraid you don't understand the problem.
I don't want to output the noise texture. Make a sprite with the godot icon, the godot icon is supposed to be shown. Now, add a shader. In that shader, add a scrolling noise texture. That shouldn't be possible, since you can only repeat, whatever is in the texture of the sprite2d Node.

But if you say to change the texture of the sprite2d node to a noise, and in the shader to set the color output to the desired texture, in this case the godot icon, the question arises, as to how one can make 2 scrolling noise textures, which one wants to combine, to get a desired effect.

**Comment by kleonc** [comment]
> @kleonc I'm afraid you don't understand the problem. I don't want to output the noise texture. Make a sprite with the godot icon, the godot icon is supposed to be shown. Now, add a shader. In that shader, add a scrolling noise texture. That shouldn't be possible, since you can only repeat, whatever is in the texture of the sprite2d Node.
> 
> But if you say to change the texture of the sprite2d node to a noise, and in the shader to set the color output to the desired texture, in this case the godot icon, the question arises, as to how one can make 2 scrolling noise textures, which one wants to combine, to get a desired effect.

Something like this?

![ipKefBLyuV](https://user-images.githubusercontent.com/9283098/233778048-997cd4d3-9ba4-4ed9-b29c-581d5a9dc32f.png)
![Godot_v4 0 2-stable_win64_TJI238wbgW](https://user-images.githubusercontent.com/9283098/233778050-8be79458-11fe-4cad-8196-daaf250f25ef.png) ![byx2NN5hys](https://user-images.githubusercontent.com/9283098/233778088-f69f363a-172c-4a3b-a9ac-88ebeb0fb919.png)

**Comment by DasGandlaf** [comment]
Oh.. so SamplerPort is a shader parameter, that takes a texture. I thought it uses the texture from the sprite2d node. I did not understand that until now. Mabe I didn't know it because of my limited knowledge of shaders, but I think maybe it should be more intuitive? To somehow mark nodes that take shader parameters. I think some people may get confused about that.
@kleonc @Chaosus Thank you both.

**Comment by DasGandlaf** [comment]
Oh.. so SamplerPort is a shader parameter, that takes a texture. I thought it uses the texture from the sprite2d node. I did not understand that until now. Mabe I didn't know it because of my limited knowledge of shaders, but I think maybe it should be more intuitive? To somehow mark nodes that take shader parameters. I think some people may get confused about that.
@kleonc @Chaosus Thank you both.

**Comment by Zireael07** [comment]
Sampler port is the literal equivalent of sampler2D in shader code, how could it be more intuitive??

**Comment by DasGandlaf** [comment]
Nobody needs your stupid question. The reason why I and many other people are using visual shader is, guess what, **because we don't know how to code shaders**. Now if you think about it, that would mean, we don't know know what Samplerport means in shader code either (mind blown).

How could it be more intuitive? Well. Just add a little text at the top, (shader parameter). Do you see, how it can possibly be more intuitive? Either don't try to make me look stupid, or reformulate your sentence.

**Comment by Zireael07** [comment]
All ports are shader parameters. So that little text is NOT helpful at all.

Things like this are exactly why I asked, because things that you think are helpful aren't necessarily helpful AND/OR as a coder I don't know what visual shader users expect.

Also: we have a code of conduct in this repo and calling others stupid does not jive with it.

**Comment by aldocd4** [comment]
Hello, I had same issue, Texture2DParameter seems to fix the problem.

But I think it would help a lot if we had possibility to set a default value in `Texture2DParameter` node (like for other parameters nodes). It's hard to switch from fullscreen visual shader editor to game window each time we update something. 

If we could see the result directly in the visual shader editor it would be cool:

![image](https://github.com/godotengine/godot/assets/2823944/a0523f6d-0e05-4202-a06e-f7c7dea4f10a)

**Comment by sdrib** [comment]
I ran in this exact same issue and first found this thread: https://github.com/godotengine/godot/issues/74310

I think that is the same issue? `Texture2DParameter` is a decent enough workaround for now but I was trying to make it work using a texture in the editor ([like the original report here](https://github.com/godotengine/godot/issues/71996#issue-1555595278)).

Is that how it eventually will work? Or would we always have to create parameters to make 2D textures repeat 🤔 ?

**Comment by sdrib** [comment]
Learning more about visual shaders in Godot and there's a gap there in documentation I guess, though it is being discussed in blog post like this one https://godotengine.org/article/major-update-visual-shaders-godot-3-2-part-2/#sampler-type

That's 3.2 so it doesn't have the 4.x  repeat option, but the 4.1 docs don't even mention sampler type (in meaningful ways) https://docs.godotengine.org/en/stable/search.html?q=sampler+type&check_keywords=yes&area=default

**Comment by ducklin5** [comment]
I've also run into this issue and wasn't sure how to get the desired result in visual shader until I read this thread.
I started with a design very similar to the first one: 
![image](https://github.com/godotengine/godot/assets/7690321/160b6b01-7b97-4ab8-9043-5f787ab855b9)

This doesn't do what it should do because I've set seamless as true but the change isn't reflected.

As for the workaround using Sampler2D, I somewhat agree that it's not very intuitive. Without this thread, I wouldn't know what that sampler2d input is for. It might be worth renaming or captioning that input to something like "External Texture".

**Comment by polymatheddy** [comment]
Has anyone had any updates on this issue? I've tried everything and can't get a noise texture to pan along with time, seems like a pretty basic thing to do, am I missing something?
